service/src/main/kotlin/fi/espoo/evaka/invoicing/controller/FinanceBasicsController.kt

```diff
@@ -349,6 +381,60 @@ WHERE id = ${bind(id)}
         }
         .execute()
 
+fun Database.Read.getServiceNeedVoucherValuesByVoucherValueRangeId(
+    voucherValueId: ServiceNeedOptionVoucherValueId
+): List<ServiceNeedOptionVoucherValueRange> =
+    createQuery {
+            sql(
+                """
+SELECT
+    id,
+    service_need_option_id,
+    validity as range,
+    base_value,
+    coefficient,
+    value,
+    base_value_under_3y,
+    coefficient_under_3y,
+    value_under_3y
+FROM service_need_option_voucher_value
+WHERE service_need_option_id = (
+  SELECT service_need_option_id
+  FROM service_need_option_voucher_value
+  WHERE id = ${bind(voucherValueId)}
+)
+ORDER by upper(validity) DESC
+"""
+            )
+        }
+        .toList<ServiceNeedOptionVoucherValueRange>()
+
+fun Database.Transaction.deleteVoucherValue(id: ServiceNeedOptionVoucherValueId) {
+    createUpdate {
+            sql(
+                """
+                DELETE
+                FROM service_need_option_voucher_value
+                WHERE id = ${bind(id)}
+            """
+            )
+        }
+        .execute()
+}
+
+fun Database.Transaction.reopenVoucherValueValidityRange(id: ServiceNeedOptionVoucherValueId) {
+    createUpdate {
```

Comment:
just FYI: you can now replace `createUpdate` with `execute` which lets you avoid a common bug of forgetting to call `.execute()` at the end